### DISCUSSION TITLE
Error 502 websocket path "notifications/hub" behind HAProxy

### DISCUSSION THREAD

**Discussion by marcogiorgio** [dani-garcia/vaultwarden#2684] [Q&A] [answered]
### Discussed in https://github.com/dani-garcia/vaultwarden/discussions/2657

Hi,
I configured Vaultwarden behind an HAProxy instance deployed on my OPNsense, using the guide "HAproxy inside PfSense (by [@RichardMawdsley](https://github.com/RichardMawdsley))". The redirects look correct, but I get a "**502 Bad Gateway
The server returned an invalid or incomplete response**" error when I go to vaultwarden.domain.net/notifications/hub.

This is the relevant HAProxy config:

acl acl01 path_beg -i /notifications/hub
acl acl02 hdr_sub(host) -i vaultwarden.domain.net
acl acl03 path_beg -i /notifications/hub/negotiate
acl acl04 path_beg -i /notifications/hub
acl acl05 path_beg -i /notifications/hub/negotiate

use_backend Vaultwarden if !acl01 acl02
use_backend Vaultwarden if acl03 acl02
use_backend Vaultwarden_Websocket if acl04 acl02
use_backend Vaultwarden_Websocket if !acl05 acl02

backend Vaultwarden_Websocket
    mode http
    balance source
    stick-table type ip size 50k expire 30m  
    stick on src
    timeout connect 30s
    timeout server 30s
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-reuse safe
    server Vaultwarden_Websocket X.X.X.X:3012

backend Vaultwarden
    mode http
    balance source
    stick-table type ip size 50k expire 30m  
    stick on src
    timeout connect 30s
    timeout server 30s
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-reuse safe
    server Vaultwarden X.X.X.X:9070 

Can anyone help? Thanks

**Comment by marcogiorgio** [comment, marked as answer]
Actually I found out that the websocket does work properly. I believed that it didn't because I followed the first method written here to test it: [Test websocket](https://vaultwarden.discourse.group/t/is-there-a-way-to-test-livesync-websockets-is-working/1435/3) Since I didn't see the message "WebSocket Protocol Error: Unable to parse WebSocket key" I thought it wasn't working. I then tried to open the debug console on my browser and I saw: "Information: WebSocket connected to wss://vaultwarden.domain.net/notifications/hub?".

**Comment by sparkssssssss** [comment]
today i update docker image to the latest ,it give me the same error code 502 when access to /notifications/hub.
while i use docker image 1.25.0,it works fine,so i have to stay the old version.

**Reply by BlackDex** [reply]
I actually think you need to check your reverse proxy config instead of staying at an older version.

**Reply by sparkssssssss** [reply]
use the 1.25.0 ,open /notifications/hub,then get the response with "WebSocket Protocol Error: Unable to parse WebSocket key.",use the latest version,i get 502 error code.

1.25.0 application log show 
```[parity_ws::io][INFO] Accepted a new tcp connection from 172.25.0.1:37154.```

the latest show:
```[vaultwarden::api::notifications][INFO] Accepting WS connection from 172.25.0.1:38506```


i follow the proxy template below,
```
    location / {
      proxy_http_version 1.1;
      proxy_set_header "Connection" "";

      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_pass http://vaultwarden-default;
    }

    location /notifications/hub/negotiate {
      proxy_http_version 1.1;
      proxy_set_header "Connection" "";

      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_pass http://vaultwarden-default;
    }

    location /notifications/hub {
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";

      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header Forwarded $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_pass http://vaultwarden-ws;
    }
```

**Reply by BlackDex** [reply]
So, both versions of Vaultwarden say they excepted the connection. Looked like Vaultwarden thinks it's ok. So with the info i have now i can only think off something else in between.

Do you have any WAF configured like ModSecurity or something. Other global config items in nginx which could cause an issue?

Because Vaultwarden isn't returning a 502. Check the nginx logs for other items.

**Reply by sparkssssssss** [reply]
i use bitwarden in my private lan network the most time. there is no waf or firewall.
plz ignore docker container timezonedifference with os


my nginx.conf
````

  location / {
    proxy_pass http://bitwarden_default;
    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }

  #location ~ ^/admin { return 404;}
  
  location /notifications/hub {
    proxy_pass http://bitwarden_ws;
    proxy_http_version 1.1;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }
  
  location /notifications/hub/negotiate {
    proxy_pass http://bitwarden_default;

    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }


}
````

nginx access & error log
````
192.168.123.150 - - [08/Sep/2022:21:11:23 +0800] "GET /notifications/hub HTTP/1.1" 502 166 "-" "curl/7.58.0" "-""192.168.123.166:63012" "502" "0.004"
2022/09/08 21:11:23 [error] 26918#26918: *2139 upstream prematurely closed connection while reading response header from upstream, client: 192.168.123.150, server: mydomain, request: "GET /notifications/hub HTTP/1.1", upstream: "http://192.168.123.166:63012/notifications/hub", host: "mydomain"
````

bitwarden log :
````
[2022-09-08 13:11:23.051][vaultwarden::api::notifications][INFO] Accepting WS connection from 192.168.123.150:43552
````


1.25.0 bitwarden log:
````
[2022-09-08 13:23:36.553][parity_ws::io][INFO] Accepted a new tcp connection from 192.168.123.150:55418.
[2022-09-08 13:23:36.553][parity_ws::handler][ERROR] WS Error <Protocol>: Unable to parse WebSocket key.
````

the totally same nginx confiuration with the 1.25.0 vaultwarden,it works perfectly,normally,vaultwarden accept the  test ws connection,then should return 400 error ,but the latest version give no response.

**Reply by BlackDex** [reply]
I still think it is an nginx issue.
Since, I use that exact same config (Since where you got it from is my example), and that works just fine for me.

Ill see if i can find something which could cause this on the nginx side.

**Reply by BlackDex** [reply]
@sparkssssssss , could you check your nginx config and see if there is any config set which mentions something like `buffer` or `timeout` which could have effect? I think it must be something regarding that.

**Reply by BlackDex** [reply]
Also, please make sure your `proxy_pass http://bitwarden_ws;` or `proxy_pass http://vaultwarden-ws;` points to the correct websocket endpoint. Default it is port 3012, but if you didn't exposed that port, then it will also fail. Not sure why that would be any different between those versions, but just be sure to check those anyway.

**Reply by sparkssssssss** [reply]
port mapping should be ok:
````
version: '3.3'
  
services:
 vaultwarden:
    image: vaultwarden/server:latest
    container_name: bitwarden
    restart: always
    volumes:
      - /vw-data:/data/
    ports:
      - 60080:80
      - 63012:3012
    environment:
      WEBSOCKET_ENABLED: "true"
      SIGNUPS_ALLOWED: "false" 
      WEB_VAULT_ENABLED: "false" 
      ADMIN_TOKEN: "xxx"
````

mentioned point in the configurateion:
proxy_buffer setting in bitwarden server section,remove buffer things with no luck.
````
upstream bitwarden_default {
    server 192.168.123.166:60080;
    keepalive 1000;
}

upstream bitwarden_ws {
   server 192.168.123.166:63012;
    keepalive 1000;
}

server {
        listen 80;
        listen 443 ssl;
        server_name domain;


        if ($scheme != "https") {
            return 301 https://$host$request_uri ;
        }

    
        proxy_buffering on;
        proxy_buffer_size 32k; 
        proxy_buffers 4 32k; 
        proxy_busy_buffers_size 64k; 
        proxy_temp_file_write_size 64k;
    
        ssl_certificate_key  /etc/nginx/ssl/xxx.key  ;
        ssl_certificate  /etc/nginx/ssl/xxx.pem;
        ssl_protocols    TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
        ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
        ssl_prefer_server_ciphers on;
        ssl_session_cache shared:SSL:10m;
        ssl_session_timeout 1d;
        resolver_timeout 1s;
        add_header Strict-Transport-Security "max-age=31536000";
        add_header Content-Security-Policy upgrade-insecure-requests;

  location / {
    proxy_pass http://bitwarden_default;
    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }

  #location ~ ^/admin { return 404;}

  
  location /notifications/hub {
    proxy_pass http://bitwarden_ws;
    proxy_http_version 1.1;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }
  
  location /notifications/hub/negotiate {
    proxy_pass http://bitwarden_default;

    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }


}
````


timeout setting in http section:
````
user root;
worker_processes auto;
pid /run/nginx.pid;
include /etc/nginx/modules-enabled/*.conf;

events {
        use epoll;
        worker_connections 65535;
}

http {


        sendfile on;
        tcp_nopush on;
        tcp_nodelay on;

        client_header_timeout 120s;
        client_body_timeout 120s;
        proxy_connect_timeout 120s;
        proxy_send_timeout 120s;
        proxy_read_timeout 120s;
        keepalive_timeout  120s 120s;
        keepalive_requests 10000;

        types_hash_max_size 2048;
        server_tokens off;


        include /etc/nginx/mime.types;
        default_type application/octet-stream;


        ssl_prefer_server_ciphers on;

        log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                  '$status $body_bytes_sent "$http_referer" '
                  '"$http_user_agent" "$http_x_forwarded_for"'
                  '"$upstream_addr" "$upstream_status" "$upstream_response_time"';

        access_log /var/log/nginx/access.log main;
        error_log /var/log/nginx/error.log error;


        gzip on;



        include /etc/nginx/conf.d/*.conf;
        include /etc/nginx/sites-enabled/*;
        proxy_cache_path /home/www/cache levels=1:2 keys_zone=info:100m max_size=200m inactive=7d use_temp_path=off;
}
````

**Reply by BlackDex** [reply]
Oke, i think i see the issue. And some other items i think should be changed.

You have a very high setting for the keepalive. This could exhaust your tcp connections, since that number is per worker, and not for the whole nginx instance. I would suggest to lower it drastically to something like `32` or `64` maybe. Also see http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive

And i see you have buffering enabled. This should be disabled for the websockets connection endpoint by adding `proxy_buffering off;`.

Please try to change those items and see what happens.

**Reply by sparkssssssss** [reply]
current bitwarden.conf

````
upstream bitwarden_default {
    server 192.168.123.166:60080;
    keepalive 32;
}

upstream bitwarden_ws {
    server 192.168.123.166:63012;
    keepalive 32;
}

server {
        listen 80;
        listen 443 ssl;
        server_name choushachou.112114.xyz;


        if ($scheme != "https") {
            return 301 https://$host$request_uri ;
        }

    
        proxy_buffering on;
        proxy_buffer_size 32k; 
        proxy_buffers 4 32k; 
        proxy_busy_buffers_size 64k; 
        proxy_temp_file_write_size 64k;
    
        ssl_certificate_key  /etc/nginx/ssl/choushachou.key  ;
        ssl_certificate  /etc/nginx/ssl/choushachou.pem;
        ssl_protocols    TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
        ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
        ssl_prefer_server_ciphers on;
        ssl_session_cache shared:SSL:10m;
        ssl_session_timeout 1d;
        resolver_timeout 1s;
        add_header Strict-Transport-Security "max-age=31536000";
        add_header Content-Security-Policy upgrade-insecure-requests;

  location / {
    proxy_pass http://bitwarden_default;
    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    #proxy_set_header Host bitwarden_default;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }

  #location ~ ^/admin { return 404;}

 
  location /notifications/hub {
    proxy_pass http://bitwarden_ws;
    proxy_buffering off;
    proxy_http_version 1.1;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
  }
  
  location /notifications/hub/negotiate {
    proxy_pass http://bitwarden_default;

    proxy_http_version 1.1;
    proxy_set_header "Connection" "";
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
  }


}
````

it doesn't work after changing nginx configuration ,thanks for your help,i decide to stay the old version ,and wait for  the next version to see what happens,thanks again.

**Reply by BlackDex** [reply]
The next version will not help, since there will be no changes to the websocket code.
There still is something incorrect in your reverse proxy code, or there is something else in between which is causing an issue.

**Reply by BlackDex** [reply]
Also, I'm looking at your log message again, it looks like you are trying to connect to with with curl, is that correct? If that is the case, that will never ever work.

It will only work via a browser, and it will only work after you are logged in.
If you start a developer console via F12 you should see a websocket `101` connection response. That should do the job and work just fine.

Other then not using it the correct way, I'm out of options here. I use nginx my self with the latest version of Vaultwarden, and it works just as it should. So, waiting for a newer version on your side will not help at all.

**Reply by BlackDex** [reply]
Also, which version of nginx are you using? Maybe there is some strange bug in an older version? Other then that, I'm out of options.

**Reply by marcogiorgio** [reply]
> Also, I'm looking at your log message again, it looks like you are trying to connect to with with curl, is that correct? If that is the case, that will never ever work.
> 
> It will only work via a browser, and it will only work after you are logged in. If you start a developer console via F12 you should see a websocket `101` connection response. That should do the job and work just fine.
> 
> Other then not using it the correct way, I'm out of options here. I use nginx my self with the latest version of Vaultwarden, and it works just as it should. So, waiting for a newer version on your side will not help at all.

@sparkssssssss check what BlackDex said. I thought that websocket didn't work because /notifications/hub threw a 502 but in the dev console I see: "**Information: WebSocket connected to wss://vaultwarden.domain.com**"

**Reply by BlackDex** [reply]
Ill see if i can some how create a test within the admin diagnostics interface to help and check if this is configured correctly.
Not sure when though.

**Reply by sparkssssssss** [reply]
I know curl will not fuction ,just for simple testing.
Normal situation ,i could get response with error code 400 instead of 502.
nginx version is 1.14.0,i will upgade to newer one to test.

**Reply by sparkssssssss** [reply]
After my testing,i think the latest version change the response code when receive incorrect ws traffic.
The real ws connection access to server,it will perform correctly,while wrong traffic comes,it will give error code 502 instead of 400,unlike the old version.

**Reply by BlackDex** [reply]
@sparkssssssss, Vaultwarden doesn't return a 502, that is nginx which does that, if you do a direct connection to the 3012 port it will never return a 502.

Also, if you really want to test the websocket connection via curl, you need to use the following example:

```bash
curl -v \
  --http1.1 \
  --include \
  --no-buffer \
  --header "Connection: Upgrade" \
  --header "Upgrade: websocket" \
  --header "Host: my.domain.tld" \
  --header "Origin: https://my.domain.tld" \
  --header "Sec-WebSocket-Key: VmF1bHR3YXJkZW4=" \
  --header "Sec-WebSocket-Version: 13" \
  "https://my.domain.tld/notifications/hub?access_token=VmF1bHR3YXJkZW4="
```

Note the `--http1.1` there, since websockets only support that protocol.
This will return a `401`, because the `access_token` and `WebSocket-Key` are invalid of course.

Anyway, this behavior is probably not going to change in any future versions, so keeping your self at an older version is not the solution, since this is just how websockets work, and as mentioned before, maybe the newer implementation is a bit more strict, that doesn't make it broken. The fact is, the websockets work just fine when they are used how they are supposed to be used.

**Reply by sparkssssssss** [reply]
Following your instruction,i verified my guess.
So,it works correctly,just the response mislead me.
Thanks for your work:)
````
root@n1:/etc/nginx/conf.d# curl -k   --include   --no-buffer   --header "Connection: Upgrade"   --header "Upgrade: websocket"   --header "Host: mydomain"   --header "Origin: https://mydomain"   --header "Sec-WebSocket-Key: VmF1bHR3YXJkZW4="   --header "Sec-WebSocket-Version: 13"   "https://mydomain/notifications/hub" 
HTTP/1.1 401 Unauthorized
Server: nginx
Date: Fri, 09 Sep 2022 13:18:53 GMT
Transfer-Encoding: chunked
Connection: keep-alive
Keep-Alive: timeout=120

root@n1:/etc/nginx/conf.d# curl -k   --include   --no-buffer  --header "Host: mydomain"   --header "Origin: https://mydomain"   --header "Sec-WebSocket-Key: VmF1bHR3YXJkZW4="   --header "Sec-WebSocket-Version: 13"   "https://mydomain/notifications/hub"              
HTTP/1.1 502 Bad Gateway
Server: nginx
Date: Fri, 09 Sep 2022 13:19:03 GMT
Content-Type: text/html
Content-Length: 166
Connection: keep-alive
Keep-Alive: timeout=120

<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
````
